### PR TITLE
disk: add WriteAt

### PR DESCRIPTION
**Pull request by yichengq**

### CODE REVIEW

File: disk/disk.go [hunk 1]
```diff
@@ -1 +1,97 @@
...
+	"io"
+	"log"
+	"os"
+)
+
+// WriteAt writes len(p) bytes from p to the file at path at offset off.
+// It returns the number of bytes written from p (0 <= n <= len(p)) and
+// any error encountered that caused the write to stop early.
+// WriteAt must return a non-nil error if it returns n < len(p).
+// TODO(xiang90): []byte or io.reader?
```
Comment (xiang90) [inline disk/disk.go]: remove this todo please

File: disk/disk.go [hunk 2]
```diff
@@ -1 +1,97 @@
 package disk
+
+import (
+	"io"
+	"log"
+	"os"
+)
+
+// WriteAt writes len(p) bytes from p to the file at path at offset off.
+// It returns the number of bytes written from p (0 <= n <= len(p)) and
```
Comment (xiang90) [inline disk/disk.go]: can we simply copy what file.WriteAt says?

File: disk/disk.go [hunk 3]
```diff
@@ -1 +1,97 @@
...
+	n := 0
+	// index that ends writing
+	end := off + int64(len(p))
+	rbuf := make([]byte, psize)
+	for i < end {
+		pidx := i / psize
+		poff := i - pidx*psize
+		wbuf := make([]byte, 0)
+
+		// read out the head of data in block
```
Comment (xiang90) [inline disk/disk.go]: well... i would 
1. find the index of the first block to write
2. find the index of the last block to write
3. if both first,last is in flen -> a
4. if first is in flen, but last is not -> b
5. if first is not in flen -> c

and  i would prefer to handle first,last out of for loop.

**Comment by yichengq** [comment]
address the comments.

@xiang90 I tried the solution you mentioned, and i found it does make logic clear, and it needs helper func to write(it needs to handle first/last block special case). Later I found that 1. represent the data that needs to be written 2. write data while handle first, last out of for loop seems more clear.

**Comment by yichengq** [comment]
@xiang90 use fillBlock func to remove duplicate code and make it simpler.

@xiang90 @yunxing @yutongp 
If WriteAt writes 10MB data starting from 10MB offset at empty file, and it fails after writing 5MB all-0 block into the file, what `n` should it return? @yunxing How do you make this case in local file system last time?

File: disk/disk.go [hunk 4]
```diff
@@ -1 +1,98 @@
...
+// WriteAt writes len(p) bytes to the File starting at byte offset off.
+// It returns the number of bytes written and an error, if any. WriteAt
+// returns a non-nil error when n != len(p).
+func WriteAt(path string, p []byte, off int64) (int, error) {
+	// block size
+	bsize := blockSize(path)
+	// payload size
+	psize := bsize - crc32Len
+
+	f, err := os.OpenFile(path, os.O_WRONLY, 0600)
```
Comment (xiang90) [inline disk/disk.go:18]: you still need to close the f.

File: disk/disk.go [hunk 5]
```diff
@@ -1 +1,98 @@
...
+		st, data = off, p
+	} else {
+		st, data = flen, append(make([]byte, off-flen), p...)
+	}
+	// index that ends writing
+	end := st + int64(len(data))
+
+	n := 0
+	stIdx, endIdx := st/psize, end/psize
+	for i := stIdx; i < endIdx; i++ {
```
Comment (xiang90) [inline disk/disk.go:60]: well... i still think no branch in for loop would be much clearer.

**Comment by yunxing** [comment]
@yichengq If it failed at writing 0 bytes, 0 should be returned since no meaningful user data has been written yet.

**Comment by yichengq** [comment]
all comments are addressed.

File: disk/disk.go [hunk 6]
```diff
@@ -1 +1,121 @@
...
+
+func getFileLength(f *os.File) int64 {
+	fi, err := f.Stat()
+	if err != nil {
+		return 0
+	}
+	return fi.Size()
+}
+
+func max(a, b int) int {
```
Comment (xiang90) [inline disk/disk.go:116]: can we move this util.go?

File: disk/disk.go [hunk 7]
```diff
@@ -1 +1,121 @@
...
+	case n < offset+int64(len(data)):
+		buf = append(buf[:offset], data...)
+	default:
+		copy(buf[offset:offset+int64(len(data))], data)
+	}
+	return writeBlock(f, index, bsize, buf)
+}
+
+// blockSize returns the block size of the file at given path.
+func blockSize(path string) int64 {
```
Comment (xiang90) [inline disk/disk.go:103]: move this one to util.go? blockSize -> getBlockSize

File: disk/disk.go [hunk 8]
```diff
@@ -1 +1,121 @@
...
+	// TODO (xiang90): implement it
+	return 4096
+}
+
+func getFileLength(f *os.File) int64 {
```
Comment (xiang90) [inline disk/disk.go:108]: move this one to util.go?

**Pull request merged**